File: dotnet/Zirve.Sdk/Queue/QueueManager.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Queue;

public class QueueManager
{
    private readonly HttpClient _httpClient;
    private readonly string _apiBase;
    private readonly string _authHeader;

    public QueueManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("queue");
        var host = cfg.GetValueOrDefault("host", "rabbitmq.zirve-infra.svc.cluster.local");
        var apiPort = cfg.GetValueOrDefault("api_port", "15672");
        
        _apiBase = $"http://{host}:{apiPort}/api";

        var user = cfg.GetValueOrDefault("user", "guest");
        var pass = cfg.GetValueOrDefault("password", "guest");
        
        var rawAuth = System.Text.Encoding.ASCII.GetBytes($"{user}:{pass}");
        _authHeader = "Basic " + Convert.ToBase64String(rawAuth);
    }

    public async Task<bool> CreateVhostAsync(string vhost)
    {
        var request = new HttpRequestMessage(HttpMethod.Put, $"{_apiBase}/vhosts/{Uri.EscapeDataString(vhost)}");
        request.Headers.Add("Authorization", _authHeader);
        
        var response = await _httpClient.SendAsync(request);
        return response.IsSuccessStatusCode;
    }

    public async Task<bool> PublishAsync(string vhost, string exchange, string routingKey, object payload)
    {
        var safeVhost = Uri.EscapeDataString(vhost);
        var safeExchange = Uri.EscapeDataString(string.IsNullOrEmpty(exchange) ? "amq.default" : exchange);
        
        var request = new HttpRequestMessage(HttpMethod.Post, $"{_apiBase}/exchanges/{safeVhost}/{safeExchange}/publish");
        request.Headers.Add("Authorization", _authHeader);
        
        var isString = payload is string;
        var reqPayload = new
        {
            properties = new { },
            routing_key = routingKey,
            payload = isString ? payload.ToString() : JsonSerializer.Serialize(payload),
            payload_encoding = "string"
        };
        
        request.Content = new StringContent(JsonSerializer.Serialize(reqPayload), System.Text.Encoding.UTF8, "application/json");

        var response = await _httpClient.SendAsync(request);
        if (!response.IsSuccessStatusCode) return false;

        using var jsonDoc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        return jsonDoc.RootElement.GetProperty("routed").GetBoolean();
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var request = new HttpRequestMessage(HttpMethod.Get, $"{_apiBase}/overview");
            request.Headers.Add("Authorization", _authHeader);
            
            var response = await _httpClient.SendAsync(request);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: typescript/src/Queue/QueueManager.ts
/**
 * Zirve Queue Manager — RabbitMQ Integration.
 *
 * Interacts with RabbitMQ Management HTTP API for generic queue controls,
 * and handles Dapr pub/sub wrappers.
 */
export class QueueManager {
  private config: Record<string, any>;
  private apiBase: string;
  private authHeader: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    const host = config.host || 'rabbitmq.zirve-infra.svc.cluster.local';
    const apiPort = config.api_port || 15672;
    this.apiBase = `http://${host}:${apiPort}/api`;

    const user = config.user || 'guest';
    const pass = config.password || 'guest';
    this.authHeader = 'Basic ' + Buffer.from(`${user}:${pass}`).toString('base64');
  }

  /**
   * Create a RabbitMQ vhost.
   */
  public async createVhost(vhost: string): Promise<boolean> {
    const safeVhost = encodeURIComponent(vhost);
    const response = await fetch(`${this.apiBase}/vhosts/${safeVhost}`, {
      method: 'PUT',
      headers: { 'Authorization': this.authHeader },
    });
    return response.ok || response.status === 204 || response.status === 201;
  }

  /**
   * Get messages from a queue via Management API (for testing/inspection only, not for high throughput).
   */
  public async getMessages(vhost: string, queue: string, count: number = 1): Promise<any[]> {
    const safeVhost = encodeURIComponent(vhost);
    const safeQueue = encodeURIComponent(queue);

    const response = await fetch(`${this.apiBase}/queues/${safeVhost}/${safeQueue}/get`, {
      method: 'POST',
      headers: {
        'Authorization': this.authHeader,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        count,
        ackmode: 'ack_requeue_true', // fetch without acknowledging
        encoding: 'auto',
      }),
    });

    if (!response.ok) {
      throw new Error(`RabbitMQ get messages failed: ${response.statusText}`);
    }

    return await response.json();
  }

  /**
   * Publish a message directly using RabbitMQ HTTP API.
   * Note: For app pub/sub, it's recommended to use Dapr sidecar instead.
   */
  public async publish(vhost: string, exchange: string, routingKey: string, payload: any): Promise<boolean> {
    const safeVhost = encodeURIComponent(vhost);
    const safeExchange = encodeURIComponent(exchange === '' ? 'amq.default' : exchange);

    const response = await fetch(`${this.apiBase}/exchanges/${safeVhost}/${safeExchange}/publish`, {
      method: 'POST',
      headers: {
        'Authorization': this.authHeader,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        properties: {},
        routing_key: routingKey,
        payload: typeof payload === 'string' ? payload : JSON.stringify(payload),
        payload_encoding: 'string'
      }),
    });

    if (!response.ok) return false;
    const data = await response.json();
    return data.routed === true;
  }

  /**
   * Basic health check for RabbitMQ Management.
   */
  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.apiBase}/overview`, {
        headers: { 'Authorization': this.authHeader },
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}
